### PR TITLE
ENH: adding casting option to numpy.stack.

### PR DESCRIPTION
**Pull request by jhonatancunha**
See #20959

np.concatenate and np.stack are similar methods, but only np.concatenate has the casting option.

This PR puts the casting option into the np.stack method to control what kind of data casting may occur

### CODE REVIEW

File: numpy/core/shape_base.py
Comment (seberg) [inline numpy/core/shape_base.py:348]: This and everything else should be keyword argument only, so `, *, casting=None`.

File: numpy/core/shape_base.pyi
Comment (seberg) [inline numpy/core/shape_base.pyi:12]: Can you maybe undo inserting these newlines?

**Review by seberg — COMMENTED** [review]
Thanks, great start.   If we add `casting=`, maybe we should probably also add `dtype=` to match `concatenate.

You accidentally checked in a generated C file, might be missing from `.gitignore`...  Please make sure to not include it.

Further we definitely need some basic tests here (pretty basic should be fine, since it just forwards to concatenate).

File: numpy/core/src/umath/loops_modulo.dispatch.c
Comment (BvB93) [inline numpy/core/src/umath/loops_modulo.dispatch.c:8]: Seems there is an auto generated file here, can you remove it?

File: numpy/core/shape_base.py
Comment (jhonatancunha) [reply numpy/core/shape_base.py]: Hello @seberg we added the option casting in the recent commits

File: numpy/core/src/umath/loops_modulo.dispatch.c
Comment (jhonatancunha) [reply numpy/core/src/umath/loops_modulo.dispatch.c]: Hello @BvB93, thanks for your reply. We removed this file in the recent commits...

**Review by seberg — COMMENTED** [review]
@jhonatancunha we still need basic tests and those style changes undone in the `pyi` file (even if they are auto-inserted maybe, my guess is the pep8 rules should not quite apply for stub files here).

Further, a very brief release note may be good (even a single bullet point), the instruction for those are in https://github.com/numpy/numpy/blob/main/doc/release/upcoming_changes/README.rst

The last thing I am wondering is if we should make sure to cover `vstack`, `hstack` etc. here, or just focus on `stack` since it is to some degree a preferred API probably.

File: numpy/core/shape_base.py
Comment (seberg) [reply numpy/core/shape_base.py:384]: The version where this will be included, which is (in current numbering) 1.24

Comment (jhonatancunha) [reply numpy/core/shape_base.py:384]: @seberg thanks for your reply, we already pushed all the commits with those changes.
We implemented some testes for the method stack with the new options, could you please review it again for us?

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (seberg) [inline doc/release/upcoming_changes/21627.new_feature.rst:2]: Line lengths of title and underline should match here (I think), I thought that sphinx would warn about it.  (Or maybe this is just a warning and doesn't matter too much?)

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (seberg) [inline doc/release/upcoming_changes/21627.new_feature.rst:3]:
```suggestion
The ``casting`` and ``dtype`` options is now available for `numpy.stack`.
```
separate casting and dtype them out (also in the title)

File: numpy/core/tests/test_shape_base.py
Comment (seberg) [inline numpy/core/tests/test_shape_base.py:459]: I would use `pytest.raises` these days, but no need to change it.

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (seberg) [reply doc/release/upcoming_changes/21627.new_feature.rst:3]: I would also suggest writing "keyword arguments"?

**Review by seberg — APPROVED** [review]
Thanks, looks good to me, the tests are probably even more thorough then necessary :).
The only thing might be polishing the release note a bit more, I think (we may also just do that as maintainers before merging though).

The one other thing is, that now that we started this, should we look into the other stacking functions as well?  That would be a good followup (unless we want to explicitly not do that).

File: numpy/core/shape_base.pyi
Comment (BvB93) [inline numpy/core/shape_base.pyi:49]:
```suggestion
    dtype: DTypeLike = ...,
```
The same applies to the overloads below, note that it will have to be imported from the `numpy._typing` namespace first.

File: numpy/core/shape_base.pyi
Comment (BvB93) [inline numpy/core/shape_base.pyi:5]:
```suggestion
from numpy import _CastingKind
```
`_CastingKind` is currently stored in the main namespace (really should move it at some point...)

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (jhonatancunha) [reply doc/release/upcoming_changes/21627.new_feature.rst]: Hello @seberg we pushed new changes, including adding these keyword arguments in hstack and vstack methods. Although 2 tests failed when pushing to GitHub, and we don't have any idea why, could you give us any tip to solve it?

File: numpy/core/shape_base.pyi
Comment (jhonatancunha) [reply numpy/core/shape_base.pyi]: Hello @BvB93 we changed the type of these keywords arguments, but 2 tests failed when we pushed to the GitHub (our local tests passed successfully). Do you have any tip to help us to solve this problem?

File: numpy/core/shape_base.pyi
Comment (BvB93) [inline numpy/core/shape_base.pyi:67]:
```suggestion
    out: None = ...,
```
@jhonatancunha I found the culprit. Seems the `out` type was accidentally changed.

Comment (jhonatancunha) [reply numpy/core/shape_base.pyi]: @BvB93 Sorry for that, I already pushed the correction for this mistake, but emerged a new error in the build process (No overload variant of "stack" matches argument types "List[List[int]]", "SubClass").

 I was looking at the overloads of concatenate method, and there are some cases that it is using DTypeLike, _DTypeLike and None on dtype keyword argument. Should I change the use of DTypeLike on every overload of stack method?

File: numpy/core/shape_base.pyi
Comment (BvB93) [reply numpy/core/shape_base.pyi:85]: Point of clarification: I'm referring to the `out` type of this specific overload.

Comment (jhonatancunha) [reply numpy/core/shape_base.pyi]: @BvB93 sorry for that, we already pushed the changes for this error. Thanks for your patience.

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (melissawm) [inline doc/release/upcoming_changes/21627.new_feature.rst:3]:
```suggestion
The ``casting`` and ``dtype`` keyword arguments are now available for `numpy.stack`. 
```

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (melissawm) [inline doc/release/upcoming_changes/21627.new_feature.rst:4]:
```suggestion
To use them, write ``np.stack(..., dtype=None, casting='same_kind')``.
```

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (melissawm) [reply doc/release/upcoming_changes/21627.new_feature.rst:3]: (Same for all release notes below)

**Review by melissawm — APPROVED** [review]
This looks good to me, thanks @jhonatancunha! I'll wait for @seberg to merge though.

File: doc/release/upcoming_changes/21627.new_feature.rst
Comment (jhonatancunha) [reply doc/release/upcoming_changes/21627.new_feature.rst]: Thanks @melissawm, we already pushed the commits with those changes!

**Comment by seberg** [comment]
Agreed, thanks everyone and Melissa for the review, lets put this in!

A few final notes, which would be nice to follow up on, but it is OK if it doesn't happen.

* We also have `dstack` (in a different file)!
* There is also `np.ma.stack` (etc.).  For me MA is in a weird state were we hope for a better replacement, but followups to fix it up would be great anyway.
* Only if we follow-up with dstack: I think it might make sense to slim the release note down to a single heading and just list all of the stacking functions individually in the text.

**Pull request merged**
